backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.h

```diff
@@ -0,0 +1,157 @@
+
+#include "backends/ebpf/ebpfObject.h"
+
+namespace EBPF {
+
+class EBPFProgram;
+
+class EBPFHashAlgorithmPSA : public EBPFObject {
+ public:
+    typedef std::vector<const IR::Expression *> argumentsList;
+
+ protected:
+    cstring baseName;
+    const EBPFProgram* program;
+    Visitor * visitor;
+
+    argumentsList unpackArguments(const IR::MethodCallExpression * expr, int dataPos);
+
+ public:
+    // keep this enum in sync with psa.p4 file
+    enum HashAlgorithm {
+        IDENTITY,
+        CRC32,
+        CRC32_CUSTOM,
+        CRC16,
+        CRC16_CUSTOM,
+        ONES_COMPLEMENT16,  // aka InternetChecksum
+        TARGET_DEFAULT
+    };
+
+    EBPFHashAlgorithmPSA(const EBPFProgram* program, cstring name)
+            : baseName(name), program(program), visitor(nullptr) {}
+
+    void setVisitor(Visitor * instance)
+    { this->visitor = instance; }
+
+    virtual unsigned getOutputWidth() const
+    { return 0; }
+
+    // decl might be a null pointer
+    virtual void emitVariables(CodeBuilder* builder, const IR::Declaration_Instance* decl) = 0;
+
+    virtual void emitClear(CodeBuilder* builder) = 0;
+    virtual void emitAddData(CodeBuilder* builder, int dataPos,
+                             const IR::MethodCallExpression * expr);
+    virtual void emitAddData(CodeBuilder* builder, const argumentsList & arguments) = 0;
+    virtual void emitGet(CodeBuilder* builder) = 0;
+
+    virtual void emitSubtractData(CodeBuilder* builder, const argumentsList & arguments) = 0;
+
+    virtual void emitGetInternalState(CodeBuilder* builder) = 0;
+    virtual void emitSetInternalState(CodeBuilder* builder,
+                                      const IR::MethodCallExpression * expr) = 0;
+};
+
+class CRCChecksumAlgorithm : public EBPFHashAlgorithmPSA {
+ protected:
+    cstring registerVar;
+    cstring initialValue;
+    cstring updateMethod;
+    cstring finalizeMethod;
+    cstring polynomial;
+    const int crcWidth;
+
+    cstring reflect(cstring str);
+
+ public:
+    CRCChecksumAlgorithm(const EBPFProgram* program, cstring name, int width)
+            : EBPFHashAlgorithmPSA(program, name), crcWidth(width) {}
+
+    unsigned getOutputWidth() const override
+    { return crcWidth; }
+
+    static void emitUpdateMethod(CodeBuilder* builder, int crcWidth);
+
+    void emitVariables(CodeBuilder* builder, const IR::Declaration_Instance* decl) override;
+
+    void emitClear(CodeBuilder* builder) override;
+    void emitAddData(CodeBuilder* builder, const argumentsList & arguments) override;
+    void emitGet(CodeBuilder* builder) override;
+
+    void emitSubtractData(CodeBuilder* builder, const argumentsList & arguments) override;
+
+    void emitGetInternalState(CodeBuilder* builder) override;
+    void emitSetInternalState(CodeBuilder* builder,
+                              const IR::MethodCallExpression * expr) override;
+};
+
+class CRC16ChecksumAlgorithm : public CRCChecksumAlgorithm {
```

Comment:
Perhaps you can add a comment describing how this protocol is used to compute the checksum: how these values are used, and what is the order that these methods are called in the generated code.